ccp.c

```diff
@@ -209,19 +225,132 @@ void ccp_connection_free(u16 sid) {
     return;
 }
 
+// lookup datapath program using program ID
+// returns  NULL on error
+struct DatapathProgram* datapath_program_lookup(u16 pid) {
+    struct DatapathProgram *prog;
+    // bounds check
+    if (pid == 0 || pid > MAX_NUM_PROGRAMS) {
+        PRINT("program index out of bounds: %d\n", pid);
+        return NULL;
+    }
+
+    prog = &datapath_programs[pid-1];
+    if (prog->index != pid) {
+        PRINT("index mismatch: pid %d, index %d", pid, prog->index);
+        return NULL;
+    }
+
+    return prog;
+
+}
+
+// scan through datapath program table for the program with this UID
+int datapath_program_lookup_uid(u32 program_uid) {
+    struct DatapathProgram *prog;
+    int i;
+    for (i=0; i < MAX_NUM_PROGRAMS; i++) {
+        prog = &datapath_programs[i];
+        if (prog->index == 0) {
+            continue;
+        }
+        if (prog->program_uid == program_uid) {
+            return (int)(prog->index);
+        }
+    }
+    return -1;
+}
+
+// saves a new datapath program into the array of datapath programs
+// returns index into datapath program array where this program is stored
+// if there is no more space, returns -1
+int datapath_program_install(struct InstallExpressionMsgHdr* install_expr_msg, char* buf) {
+    u16 pid;
+    int ok;
+    int i;
+    struct DatapathProgram* program;
+    struct InstructionMsg* current_instr;
+    char* msg_ptr; // for reading from char* buf
+    msg_ptr = buf;
+    for (pid = 0; pid < MAX_NUM_PROGRAMS; pid++) {
+        program = &datapath_programs[pid];
+        if (program->index == 0) {
+            // found a free slot
```

Comment:
not exactly related to this line, but more generally -- when are programs removed from this array? if they're not, then even if your algorithm only installs 1 program, if you run it 11 times, the 11th time there will be no free slots left. we need to know when ccp exits so that we can "free" (not actually free the memory, but release the slots)